### PR TITLE
Allow catalog.internal to be configurable (#2865)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: src/azul/__init__.py [hunk 1]
```diff
@@ -528,19 +529,16 @@ def __attrs_post_init__(self):
...
 
-        @property
-        def is_internal(self):
-            # FIXME: Remove the second term
-            #        https://github.com/DataBiosphere/azul/issues/2865
-            return self.is_integration_test_catalog or self.name == 'dcp3'
-
         @classmethod
         def from_json(cls, name: str, spec: JSON) -> 'Config.Catalog':
             plugins = {
```
Comment (jessebrennan) [inline src/azul/__init__.py:534]: This is more feedback for the parent PR, but I wanted to point out that there's no enforcement for *which* plugins are configured. In other words, `plugins = {}` would parse here as well as some made up plugins. We might want to be stricter about that.

**Review by jessebrennan — CHANGES_REQUESTED** [review]

File: src/azul/__init__.py [hunk 2]
```diff
@@ -528,19 +529,16 @@ def __attrs_post_init__(self):
         def is_integration_test_catalog(self) -> bool:
             return self._it_catalog_re.match(self.name) is not None
 
-        @property
-        def is_internal(self):
-            # FIXME: Remove the second term
-            #        https://github.com/DataBiosphere/azul/issues/2865
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/__init__.py:534]: From the issue history, it appears that the PR/FIXME were created together.
https://github.com/DataBiosphere/azul/commit/def466aef134867fda4b2a88f39373e8797137a7 

Also note that `dcp3` is not used in any deployment.

File: src/azul/__init__.py (hunk 1 above)
Comment (amarjandu) [reply src/azul/__init__.py:534]: I'll consider adding something to the parent PR.

Comment (amarjandu) [reply src/azul/__init__.py:534]: https://github.com/DataBiosphere/azul/pull/3109/files#diff-a5ab528d23a14c23dd1439c679fe8f2af5e5bcab793aee95451aeefe4f7b86dfR521

Comment (amarjandu) [reply src/azul/__init__.py:534]:
```
config.Catalog.from_json(name='hi', spec=dict(atlas='foo', plugins={}))
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 543, in from_json
    return cls(name=name, atlas=spec['atlas'], plugins=plugins)
  File "<attrs generated init azul.Config.Catalog>", line 6, in __init__
    self.__attrs_post_init__()
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 523, in __attrs_post_init__
    require(all_types == configured_types,
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 1097, in require
    reject(not condition, *args, exception=exception)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 1112, in reject
    raise exception(*args)
azul.RequirementError: ('Missing or extra plugin types', {'repository', 'metadata'})
```

File: src/azul/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/azul/__init__.py:534]: Sorry, I got mixed up when looking at the diff, everything here is (obviously) ok.

**Review by jessebrennan — APPROVED** [review]
Lookin good :sunglasses:

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/__init__.py [hunk 3]
```diff
@@ -526,21 +527,24 @@ def __attrs_post_init__(self):
...
         @cached_property
         def is_integration_test_catalog(self) -> bool:
-            return self._it_catalog_re.match(self.name) is not None
-
-        @property
-        def is_internal(self):
-            # FIXME: Remove the second term
-            #        https://github.com/DataBiosphere/azul/issues/2865
-            return self.is_integration_test_catalog or self.name == 'dcp3'
+            if self._it_catalog_re.match(self.name) is not None:
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:530]: flip `if`, eliminate `not`

File: src/azul/__init__.py [hunk 4]
```diff
@@ -526,21 +527,24 @@ def __attrs_post_init__(self):
...
+                require(self.internal,
+                        'IT catalogs must have their `internal` property set to True',
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:532]:
```suggestion
                        'IT catalogs must be internal',
```

File: src/azul/__init__.py [hunk 5]
```diff
@@ -526,21 +527,24 @@ def __attrs_post_init__(self):
...
+                        attr.asdict(self))
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:533]: Why `asdict`?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/__init__.py (hunk 5 above)
Comment (amarjandu) [reply src/azul/__init__.py:533]: Provides more context if/when this fires off.

ex. w/ self.internal in the message
```
from azul import config
config.integration_test_catalogs
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/amar/dev/hca/azul/src/azul/caching.py", line 165, in __get__
    value = obj.__dict__[self.fget.__name__] = self.fget(obj)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 584, in integration_test_catalogs
    return {
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 587, in <dictcomp>
    if catalog.is_integration_test_catalog
  File "/Users/amar/dev/hca/azul/src/azul/caching.py", line 165, in __get__
    value = obj.__dict__[self.fget.__name__] = self.fget(obj)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 531, in is_integration_test_catalog
    require(self.internal, 'IT catalogs must be internal', self.internal)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 1148, in require
    reject(not condition, *args, exception=exception)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 1163, in reject
    raise exception(*args)
azul.RequirementError: ('IT catalogs must be internal', False)
```

vs .asdict
```
from azul import config
config.integration_test_catalogs
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/amar/dev/hca/azul/src/azul/caching.py", line 165, in __get__
    value = obj.__dict__[self.fget.__name__] = self.fget(obj)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 584, in integration_test_catalogs
    return {
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 587, in <dictcomp>
    if catalog.is_integration_test_catalog
  File "/Users/amar/dev/hca/azul/src/azul/caching.py", line 165, in __get__
    value = obj.__dict__[self.fget.__name__] = self.fget(obj)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 531, in is_integration_test_catalog
    require(self.internal, 'IT catalogs must be internal', attr.asdict(self))
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 1148, in require
    reject(not condition, *args, exception=exception)
  File "/Users/amar/dev/hca/azul/src/azul/__init__.py", line 1163, in reject
    raise exception(*args)
azul.RequirementError: ('IT catalogs must be internal', {'name': 'it2ebi', 'atlas': 'hca', 'internal': False, 'plugins': {'metadata': {'name': 'hca'}, 'repository': {'name': 'tdr'}}})
```

I suppose adding `self.name, self.internal`  would also provide context, but why not give an operator all information about the catalog with the potential configuration issue?

Comment (hannes-ucsc) [reply src/azul/__init__.py]: Not what I meant. Why `asdict(self)` and not not just `self` and rely on `repr`>

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
https://github.com/DataBiosphere/azul/pull/3276#discussion_r689743656

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**